### ISSUE TITLE
Add flag for easily finetuning heads / linear probing to AutoModelforSequenceClassification

### ISSUE THREAD

**Issue by 0amp** [huggingface/transformers#27730]
### Feature request

Previous work has shown that last layer linear probing is cheaper and often generalizes better than normal finetuning (see [1](https://arxiv.org/pdf/2202.10054.pdf)). I imagine this could be implemented as a flag to AutoModelforSequenceClassification so that only the last layer classification head is trained. I believe this can be done manually by setting all the parameters except the last one to not track gradients, but a flag may be easier and encourage adoption. 

It may also be nice to have linear probing available at earlier layers (eg halfway through the model). This could be done through using the output_hidden_states flag during a forward pass. Mid layer linear probing can occasionally be more effective and is a widely used technique in the interpretability literature (see [2](https://direct.mit.edu/coli/article/48/1/207/107571/Probing-Classifiers-Promises-Shortcomings-and), [3](https://proceedings.neurips.cc/paper_files/paper/2019/file/159c1ffe5b61b41b3c4d8f4c2150f6c4-Paper.pdf), [4](https://arxiv.org/abs/2311.03658#:~:text=Informally%2C%20the%20%27linear%20representation%20hypothesis,directions%20in%20some%20representation%20space.), [5](https://arxiv.org/abs/2310.01405), and many others). 

Alternatively, if this were implemented generally for AutoModel (or maybe AutoModelforCausalLM?), it could use a wider variety of models. This feature could also be paired with an update that automatically allows models to be used for sequence classification by appending a final linear layer. 

### Motivation

Finetuning a head is extremely memory efficient and extremely fast (order of 1k parameters for most models, linear probes generally train in seconds, the main bottleneck will just be the forward pass) and oftentimes performs close to finetuning for classification tasks. It has also been shown to perform better OOD. 

### Your contribution

I can provide feedback and testing, have not looked deep enough to know how to fully implement this
